src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4526,18 +4568,62 @@ private bool HasTopLevelNullabilityConversion(TypeWithAnnotations source, TypeWi
 
             // method return type -> conversion "to" type
             // May be distinct from method return type for Nullable<T>.
-            operandType = ClassifyAndApplyConversion(operandOpt ?? node, TypeWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType), operandType,
-                useLegacyWarnings, assignmentKind, target, reportWarnings: reportRemainingWarnings, location: operandOpt is null ? location : null);
+            operandType = ClassifyAndVisitConversion(
+                convertedNode,
+                TypeWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType),
+                operandType,
+                useLegacyWarnings,
+                assignmentKind,
+                target,
+                reportWarnings: reportRemainingWarnings,
+                fromExplicitCast: false,
+                location: convertedNode is null ? diagnosticLocation : null);
 
             // conversion "to" type -> final type
-            // https://github.com/dotnet/roslyn/issues/29959 If the original conversion was
-            // explicit, this conversion should not report nested nullability mismatches.
-            // (see NullableReferenceTypesTests.ExplicitCast_UserDefined_02).
-            operandType = ClassifyAndApplyConversion(node, targetTypeWithNullability, operandType,
-                useLegacyWarnings, assignmentKind, target, reportWarnings: reportRemainingWarnings, location);
+            // We should only pass fromExplicitCast here. Given the following example:
+            //
+            // class A { public static explicit operator C(A a) => throw null!; }
+            // class C
+            // {
+            //     void M() => var c = (C?)new A();
+            // }
+            //
+            // This final conversion from the method return type "C" to the cast type "C?" is
+            // where we will need to ensure that this is counted as an explicit cast, so that
+            // the resulting operandType is nullable if that was introduced via cast.
+            operandType = ClassifyAndVisitConversion(
+                conversionOpt ?? convertedNode,
+                targetTypeWithNullability,
+                operandType,
+                useLegacyWarnings,
+                assignmentKind,
+                target,
+                reportWarnings: reportRemainingWarnings,
+                fromExplicitCast: conversionOpt?.ExplicitCastInCode ?? false,
+                diagnosticLocation);
```

Comment:
>diagnosticLocation [](start = 16, length = 18)

Why do other calls to `ClassifyAndVisitConversion` in this method pass `location: convertedNode is null ? diagnosticLocation : null` but this call simply passes `diagnosticLocation`?

Reply:
Because I apparently missed refactoring those to just use diagnostic location.

---
In reply to: [282238663](https://github.com/dotnet/roslyn/pull/35544#discussion_r282238663) [](ancestors = 282238663)

Reply:
More generally though, those are on the "operand" to the conversion, while this one is on the "result of the conversion". `diagnosticLocation` points to the latter syntax, so we pass null above to get the former.

---
In reply to: [282271624](https://github.com/dotnet/roslyn/pull/35544#discussion_r282271624) [](ancestors = 282271624,282238663)